File: Job Sequencing/Job_Sequencing.py
'''
    JOB SEQUENCING ALGORITHM

    There are n tasks(job) given to perform. Each of them has a certain profit associated 
    with them. Also, there is a deadline given to every task. The tasks must be completed
    before their deadlines.
    The problem is to find the maximum profit we can achieve while satisfying all the conditions.

'''

n = 9
deadline = [5,4,3,3,4,5,2,3,7]
profit = [85,25,16,40,55,19,92,80,15]
jobs = []
maxDeadline = 0         #Maximum deadline among the jobs.
for i in range(0,n):
    jobs.append((profit[i],(deadline[i],i+1)))
    maxDeadline = max(maxDeadline,deadline[i])
jobs.sort(key=lambda x:x[0])    #Jobs are sorted according to their profits.
time = []
for i in range(0,maxDeadline+1):
    time.append(0)
for i in range(len(jobs)-1,-1,-1):  #We iterate over the jobs in decreasing order of the profit.
    for j in range(jobs[i][1][0],0,-1):
        if time[j]==0:      #We assign the job to the first available slot before it's deadline.
            time[j]=jobs[i][1][1]
            break

maxProfit = 0
for i in range(1,maxDeadline+1):
    print(time[i],end=" ")
    if time[i]==0:
        continue
    maxProfit = maxProfit + profit[time[i]-1]
print("")
print("Max profit : "+str(maxProfit))

'''
        Output : 4 7 8 5 1 0 9
        Max profit : 367
        where time[i] denotes the number of the job which is assigned to time i.
        time[i] = 0 denotes no job assigned to that time.
'''

File: Job Sequencing/Job_Sequencing.java
import java.util.*; 
  
// A Simple Disjoint Set Data Structure 
class DisjointSet 
{ 
    int parent[]; 
  
    // Constructor 
    DisjointSet(int n) 
    { 
        parent = new int[n + 1]; 
  
        // Every node is a parent of itself 
        for (int i = 0; i <= n; i++) 
            parent[i] = i; 
    } 
  
    // Path Compression 
    int find(int s) 
    { 
        /* Make the parent of the nodes in the path 
           from u--> parent[u] point to parent[u] */
        if (s == parent[s]) 
            return s; 
        return parent[s] = find(parent[s]); 
    } 
  
    // Makes u as parent of v. 
    void merge(int u, int v) 
    { 
        //update the greatest available 
        //free slot to u 
        parent[v] = u; 
    } 
} 
  
class Job implements Comparator<Job> 
{ 
    // Each job has a unique-id, profit and deadline 
    char id; 
    int deadline, profit; 
  
    // Constructors 
    public Job() { } 
    public Job(char id,int deadline,int profit) 
    { 
        this.id = id; 
        this.deadline = deadline; 
        this.profit = profit; 
    } 
  
    // Returns the maximum deadline from the set of jobs 
    public static int findMaxDeadline(ArrayList<Job> arr) 
    { 
        int ans = Integer.MIN_VALUE; 
        for (Job temp : arr) 
            ans = Math.max(temp.deadline, ans); 
        return ans; 
    } 
  
    // Prints optimal job sequence 
    public static void printJobScheduling(ArrayList<Job> arr) 
    { 
        // Sort Jobs in descending order on the basis 
        // of their profit 
        Collections.sort(arr, new Job()); 
  
        // Find the maximum deadline among all jobs and 
        // create a disjoint set data structure with 
        // maxDeadline disjoint sets initially. 
        int maxDeadline = findMaxDeadline(arr); 
        DisjointSet dsu = new DisjointSet(maxDeadline); 
  
        // Traverse through all the jobs 
        for (Job temp : arr) 
        { 
            // Find the maximum available free slot for 
            // this job (corresponding to its deadline) 
            int availableSlot = dsu.find(temp.deadline); 
  
  
            // If maximum available free slot is greater 
            // than 0, then free slot available 
            if (availableSlot > 0) 
            { 
                // This slot is taken by this job 'i' 
                // so we need to update the greatest free 
                // slot. Note that, in merge, we make 
                // first parameter as parent of second 
                // parameter.  So future queries for 
                // availableSlot will return maximum slot 
                // from set of "availableSlot - 1" 
                dsu.merge(dsu.find(availableSlot - 1), 
                                   availableSlot); 
                System.out.print(temp.id + " "); 
            } 
        } 
        System.out.println(); 
    } 
  
    // Used to sort in descending order on the basis 
    // of profit for each job 
    public int compare(Job j1, Job j2) 
    { 
        return j1.profit > j2.profit? -1: 1; 
    } 
} 
  
// Driver code 
class Main 
{ 
    public static void main(String args[]) 
    { 
        ArrayList<Job> arr=new ArrayList<Job>(); 
        arr.add(new Job('a',2,100)); 
        arr.add(new Job('b',1,19)); 
        arr.add(new Job('c',2,27)); 
        arr.add(new Job('d',1,25)); 
        arr.add(new Job('e',3,15)); 
        System.out.println("Following jobs need to be "+ 
                           "executed for maximum profit"); 
        Job.printJobScheduling(arr); 
    } 
} 
